doc/stunnel.8
.\" Automatically generated by Pod::Man 2.1801 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\" ========================================================================
.\"
.IX Title "STUNNEL 8"
.TH STUNNEL 8 "2009.11.20" "4.29" "stunnel"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
stunnel \- universal SSL tunnel
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\fBUnix:\fR" 4
.IX Item "Unix:"
\&\fBstunnel\fR [<filename>] | \-fd n | \-help | \-version | \-sockets
.IP "\fB\s-1WIN32:\s0\fR" 4
.IX Item "WIN32:"
\&\fBstunnel\fR [ [\-install | \-uninstall | \-start | \-stop]
    [\-quiet] [<filename>] ] | \-help | \-version | \-sockets
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBstunnel\fR program is designed to work as \fI\s-1SSL\s0\fR encryption wrapper 
between remote clients and local (\fIinetd\fR\-startable) or remote
servers. The concept is that having non-SSL aware daemons running on
your system you can easily set them up to communicate with clients over
secure \s-1SSL\s0 channels.
.PP
\&\fBstunnel\fR can be used to add \s-1SSL\s0 functionality to commonly used \fIInetd\fR
daemons like \s-1POP\-2\s0, \s-1POP\-3\s0, and \s-1IMAP\s0 servers, to standalone daemons like
\&\s-1NNTP\s0, \s-1SMTP\s0 and \s-1HTTP\s0, and in tunneling \s-1PPP\s0 over network sockets without
changes to the source code.
.PP
This product includes cryptographic software written by
Eric Young (eay@cryptsoft.com)
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "<\fBfilename\fR>" 4
.IX Item "<filename>"
Use specified configuration file
.IP "\fB\-fd n\fR (Unix only)" 4
.IX Item "-fd n (Unix only)"
Read the config file from specified file descriptor
.IP "\fB\-help\fR" 4
.IX Item "-help"
Print \fBstunnel\fR help menu
.IP "\fB\-version\fR" 4
.IX Item "-version"
Print \fBstunnel\fR version and compile time defaults
.IP "\fB\-sockets\fR" 4
.IX Item "-sockets"
Print default socket options
.IP "\fB\-install\fR (\s-1NT/2000/XP\s0 only)" 4
.IX Item "-install (NT/2000/XP only)"
Install \s-1NT\s0 Service
.IP "\fB\-uninstall\fR (\s-1NT/2000/XP\s0 only)" 4
.IX Item "-uninstall (NT/2000/XP only)"
Uninstall \s-1NT\s0 Service
.IP "\fB\-start\fR (\s-1NT/2000/XP\s0 only)" 4
.IX Item "-start (NT/2000/XP only)"
Start \s-1NT\s0 Service
.IP "\fB\-stop\fR (\s-1NT/2000/XP\s0 only)" 4
.IX Item "-stop (NT/2000/XP only)"
Stop \s-1NT\s0 Service
.IP "\fB\-quiet\fR (\s-1NT/2000/XP\s0 only)" 4
.IX Item "-quiet (NT/2000/XP only)"
Don't display a message box when successfully installed or uninstalled \s-1NT\s0 service
.SH "CONFIGURATION FILE"
.IX Header "CONFIGURATION FILE"
Each line of the configuration file can be either:
.IP "\(bu" 4
an empty line (ignored)
.IP "\(bu" 4
a comment starting with ';' (ignored)
.IP "\(bu" 4
an 'option_name = option_value' pair
.IP "\(bu" 4
\&'[service_name]' indicating a start of a service definition
.SS "\s-1GLOBAL\s0 \s-1OPTIONS\s0"
.IX Subsection "GLOBAL OPTIONS"
.IP "\fBchroot\fR = directory (Unix only)" 4
.IX Item "chroot = directory (Unix only)"
directory to chroot \fBstunnel\fR process
.Sp
\&\fBchroot\fR keeps \fBstunnel\fR in chrooted jail.  \fICApath\fR, \fICRLpath\fR, \fIpid\fR
and \fIexec\fR are located inside the jail and the patches have to be relative
to the directory specified with \fBchroot\fR.
.Sp
To have libwrap (\s-1TCP\s0 Wrappers) control effective in a chrooted environment
you also have to copy its configuration files (/etc/hosts.allow and
/etc/hosts.deny) there.
.IP "\fBcompression\fR = zlib | rle" 4
.IX Item "compression = zlib | rle"
select data compression algorithm
.Sp
default: no compression
.Sp
zlib compression of OpenSSL 0.9.8 or above is not backward compatible with
OpenSSL 0.9.7.
.Sp
rle compression is currently not implemented by the OpenSSL library.
.IP "\fBdebug\fR = [facility.]level" 4
.IX Item "debug = [facility.]level"
debugging level
.Sp
Level is a one of the syslog level names or numbers
emerg (0), alert (1), crit (2), err (3), warning (4), notice (5),
info (6), or debug (7).  All logs for the specified level and
all levels numerically less than it will be shown.  Use \fBdebug = debug\fR or
\&\fBdebug = 7\fR for greatest debugging output.  The default is notice (5).
.Sp
The syslog facility 'authpriv' will be used unless a facility name is supplied.
(Facilities are not supported on Win32.)
.Sp
Case is ignored for both facilities and levels.
.IP "\fB\s-1EGD\s0\fR = egd path (Unix only)" 4
.IX Item "EGD = egd path (Unix only)"
path to Entropy Gathering Daemon socket
.Sp
Entropy Gathering Daemon socket to use to feed OpenSSL random number
generator.  (Available only if compiled with OpenSSL 0.9.5a or higher)
.IP "\fBengine\fR = auto | <engine id>" 4
.IX Item "engine = auto | <engine id>"
select hardware engine
.Sp
default: software-only cryptography
.Sp
There's an example in '\s-1EXAMPLES\s0' section.
.IP "\fBengineCtrl\fR = command[:parameter]" 4
.IX Item "engineCtrl = command[:parameter]"
control hardware engine
.Sp
Special commands \*(L"\s-1LOAD\s0\*(R" and \*(L"\s-1INIT\s0\*(R" can be used to load and initialize the
engine cryptogaphic module.
.IP "\fBfips\fR = yes | no" 4
.IX Item "fips = yes | no"
Enable or disable \s-1FIPS\s0 140\-2 mode.
.Sp
This option allows to disable entering \s-1FIPS\s0 mode if stunnel was compiled with
\&\s-1FIPS\s0 140\-2 support.
.Sp
default: yes
.IP "\fBforeground\fR = yes | no (Unix only)" 4
.IX Item "foreground = yes | no (Unix only)"
foreground mode
.Sp

src/stunnel.c
/*
 *   stunnel       Universal SSL tunnel
 *   Copyright (c) 1998-2006 Michal Trojnara <Michal.Trojnara@mirt.net>
 *                 All Rights Reserved
 *
 *   Version:      4.15             (stunnel.c)
 *   Date:         2006.03.11
 *
 *   Author:       Michal Trojnara  <Michal.Trojnara@mirt.net>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *   In addition, as a special exception, Michal Trojnara gives
 *   permission to link the code of this program with the OpenSSL
 *   library (or with modified versions of OpenSSL that use the same
 *   license as OpenSSL), and distribute linked combinations including
 *   the two.  You must obey the GNU General Public License in all
 *   respects for all of the code used other than OpenSSL.  If you modify
 *   this file, you may extend this exception to your version of the
 *   file, but you are not obligated to do so.  If you do not wish to
 *   do so, delete this exception statement from your version.
 */

#include "common.h"
#include "prototypes.h"

    /* Prototypes */
static void daemon_loop(void);
static void accept_connection(LOCAL_OPTIONS *);
static void handle_control_connection(void);
static void get_limits(void); /* setup global max_clients and max_fds */
#if !defined (USE_WIN32) && !defined (__vms)
static void drop_privileges(void);
static void daemonize(void);
static void create_pid(void);
static void delete_pid(void);
static void create_control_socket(void);
#endif

    /* Error/exceptions handling functions */
#ifndef USE_WIN32
static void signal_handler(int);
#endif

int volatile num_clients=0; /* Current number of clients */

int control_socket_fd; /* Local socket for control messages */

    /* Functions */

#ifndef USE_WIN32
int main(int argc, char* argv[]) { /* execution begins here 8-) */

    main_initialize(argc>1 ? argv[1] : NULL, argc>2 ? argv[2] : NULL);

    signal(SIGPIPE, SIG_IGN); /* avoid 'broken pipe' signal */
    if(signal(SIGTERM, SIG_IGN)!=SIG_IGN)
        signal(SIGTERM, signal_handler);
    if(signal(SIGQUIT, SIG_IGN)!=SIG_IGN)
        signal(SIGQUIT, signal_handler);
    if(signal(SIGINT, SIG_IGN)!=SIG_IGN)
        signal(SIGINT, signal_handler);
    if(signal(SIGHUP, SIG_IGN)!=SIG_IGN)
        signal(SIGHUP, signal_handler);
    /* signal(SIGSEGV, signal_handler); */

    main_execute();

    return 0; /* success */
}
#endif

CLI *all_clients;

void main_initialize(char *arg1, char *arg2) {
    ssl_init(); /* initialize SSL library */
    sthreads_init(); /* initialize critical sections & SSL callbacks */
    parse_config(arg1, arg2);
    log_open();
    stunnel_info(0);
	all_clients = NULL;
}

void main_execute(void) {
    /* check if started from inetd */
    if(local_options.next) { /* there are service sections -> daemon mode */
        daemon_loop();
    } else { /* inetd mode */
#if !defined (USE_WIN32) && !defined (__vms)
        max_fds=FD_SETSIZE; /* just in case */
        drop_privileges();
#endif
        num_clients=1;
        client(alloc_client_session(&local_options, 0, 1));
    }
    log_close();
}

static void daemon_loop(void) {
    SOCKADDR_UNION addr;
    s_poll_set fds;
    LOCAL_OPTIONS *opt;

    get_limits();
    s_poll_zero(&fds);
#ifndef USE_WIN32
    s_poll_add(&fds, signal_pipe_init(), 1, 0);
#endif

    if(!local_options.next) {
        s_log(LOG_ERR, "No connections defined in config file");
        exit(1);
    }

    num_clients=0;

    /* bind local ports */
    for(opt=local_options.next; opt; opt=opt->next) {
        if(!opt->option.accept) /* no need to bind this service */
            continue;
        memcpy(&addr, &opt->local_addr.addr[0], sizeof(SOCKADDR_UNION));
        if((opt->fd=socket(addr.sa.sa_family, SOCK_STREAM, 0))<0) {
            sockerror("local socket");
            exit(1);
        }
        if(alloc_fd(opt->fd))
            exit(1);
        if(set_socket_options(opt->fd, 0)<0)
            exit(1);
        s_ntop(opt->local_address, &addr);
        if(bind(opt->fd, &addr.sa, addr_len(addr))) {
            s_log(LOG_ERR, "Error binding %s to %s",
                opt->servname, opt->local_address);
            sockerror("bind");
            exit(1);
        }
        s_log(LOG_DEBUG, "%s bound to %s", opt->servname, opt->local_address);
        if(listen(opt->fd, 5)) {
            sockerror("listen");
            exit(1);
        }
#ifdef FD_CLOEXEC
        fcntl(opt->fd, F_SETFD, FD_CLOEXEC); /* close socket in child execvp */
#endif
        s_poll_add(&fds, opt->fd, 1, 0);
    }

#if !defined (USE_WIN32) && !defined (__vms)
    if(!(options.option.foreground))
        daemonize();
    drop_privileges();
    create_pid();
	create_control_socket();
#endif /* !defined USE_WIN32 && !defined (__vms) */

    if(options.control_socket){
	  s_log(LOG_DEBUG, "Adding control fd %d", control_socket_fd);
	  s_poll_add(&fds, control_socket_fd, 1, 0);
	}

    /* create exec+connect services */
    for(opt=local_options.next; opt; opt=opt->next) {
        if(opt->option.accept) /* skip ordinary (accepting) services */
            continue;
        enter_critical_section(CRIT_CLIENTS); /* for multi-cpu machines */
        ++num_clients;
        leave_critical_section(CRIT_CLIENTS);
        create_client(-1, -1, alloc_client_session(opt, -1, -1), client);
    }

    while(1) {
        if(s_poll_wait(&fds, -1)<0) { /* non-critical error */
            log_error(LOG_INFO, get_last_socket_error(),
                "daemon_loop: s_poll_wait");
            sleep(1); /* to avoid log trashing */
        } else {
            for(opt=local_options.next; opt; opt=opt->next)
                if(s_poll_canread(&fds, opt->fd))
                    accept_connection(opt);
			if(options.control_socket)
			  if(s_poll_canread(&fds, control_socket_fd))
				handle_control_connection();
        }
    }
    s_log(LOG_ERR, "INTERNAL ERROR: End of infinite loop 8-)");
}

static void handle_control_ping(int s){
  ssize_t n;
  char buf[1024];
  bzero(buf, sizeof(buf));
  n = snprintf(buf + sizeof(uint16_t), sizeof(buf) - sizeof(uint16_t), "PONG");
  *((uint16_t*)buf) = n;
  if ((n = send(s, buf, n + sizeof(uint16_t), 0)) < 0){
	s_log(LOG_ERR, "Failed to reply to a message on the control socket: %s", strerror(errno));
	return;
  }
}

static void handle_unknown_message(int s){
  ssize_t n;
  char buf[1024];
  bzero(buf, sizeof(buf));
  n = snprintf(buf + sizeof(uint16_t), sizeof(buf) - sizeof(uint16_t), "UNKNOWN MESSAGE");
  *((uint16_t*)buf) = n;
  if ((n = send(s, buf, n + sizeof(uint16_t), 0)) < 0){
	s_log(LOG_ERR, "Failed to reply to a message on the control socket: %s", strerror(errno));
	return;
  }
}

static int parse_ip_port(const char *x, struct sockaddr_in *sockaddr, unsigned short *port){
  char *ip;
  const char *colon;
  int r;

  /* We expect "addr" to have the form IP:port */
  colon = strchr(x, ':');
  if (!colon){
	s_log(LOG_ERR, "Failed to parse IP:port [%s]", x);
	return 1;
  }
  *port = atoi(colon + 1); /* worst case colon + 1 is the NULL terminator */

  ip = strdup(x);
  r = 0;
  ip[colon - x] = '\0';
  if (inet_pton(AF_INET, ip, &(sockaddr->sin_addr)) <= 0){
	s_log(LOG_ERR, "Failed to parse IP from argument: %s (%s)", x, ip);
	r = 2;
	goto out;
  }

  s_log(LOG_DEBUG, "IP=%s Port=%d", ip, *port);
 out:
  free(ip);
  return r;
}

static void handle_control_stat(int s, const char *addr, int len){
  unsigned short port = 0;
  struct sockaddr_in output_sockaddr; /* address of local service */
  struct sockaddr_in input_sockaddr;  /* address of original client */
  socklen_t input_sockaddr_len;
  CLI *c;
  ssize_t n;
  char buf[1024];
  bzero(buf, sizeof(buf));

  if (parse_ip_port(addr, &output_sockaddr, &port) != 0){
	s_log(LOG_ERR, "Failed to parse STAT argument: %s", addr);
	handle_unknown_message(s);
	return;
  }

  /* default if we don't find anything */
  n = snprintf(buf + sizeof(uint16_t), sizeof(buf) - sizeof(uint16_t), "UNKNOWN");

  enter_critical_section(CRIT_CLIENTS);
  c = all_clients;
  while (c != NULL){
	if (c->our_sockname_valid != 0){
	  s_log(LOG_DEBUG, "Checking port = %d (=%d)", ntohs(c->our_sockname.sin_port), port);
	  if ((memcmp(&output_sockaddr.sin_addr, &c->our_sockname.sin_addr, sizeof(output_sockaddr.sin_addr)) == 0) && (port == ntohs(c->our_sockname.sin_port))) {
		s_log(LOG_DEBUG, "Found");
		input_sockaddr_len = sizeof(input_sockaddr);
		if (getsockname(c->local_rfd.fd, (struct sockaddr*) &input_sockaddr, &input_sockaddr_len) != 0){
		  s_log(LOG_ERR, "getsockname failed: %s", strerror(errno));
		  n = snprintf(buf + sizeof(uint16_t), sizeof(buf) - sizeof(uint16_t), "ERROR");

		} else {
		  bzero(buf, sizeof(buf));
		  if (inet_ntop(AF_INET, (const void*) &input_sockaddr.sin_addr, buf + sizeof(uint16_t), sizeof(buf) - sizeof(uint16_t)) == NULL){
			s_log(LOG_ERR, "inet_ntop failed: %s", strerror(errno));
			n = snprintf(buf + sizeof(uint16_t), sizeof(buf) - sizeof(uint16_t), "ERROR");
		  } else {
			n = strlen(buf + sizeof(uint16_t)) + sizeof(uint16_t);
			n += snprintf(buf + n, sizeof(buf) - n, ":%d", ntohs(input_sockaddr.sin_port));
		  }
		  s_log(LOG_DEBUG, "returning [%s] length %d", buf + sizeof(uint16_t), n);
		}
		break;
	  }
	}
	c = c->next;
  }
  leave_critical_section(CRIT_CLIENTS);

  *((uint16_t*)buf) = n;
  if ((n = send(s, buf, n + sizeof(uint16_t), 0)) < 0){
	s_log(LOG_ERR, "Failed to reply to a message on the control socket: %s", strerror(errno)); 
  }
 
}

#define PING "ping"
#define STAT "stat"
#define SPLICE "splice"


static void handle_splice(int s, const char *addr, int len, int receive_fd){
  unsigned short port = 0;
  struct sockaddr_in output_sockaddr; /* address of local service */
  CLI *c;
  ssize_t n;
  char buf[1024];
  char control_message;

  if (parse_ip_port(addr, &output_sockaddr, &port) != 0){
	s_log(LOG_ERR, "Failed to parse SPLICE argument: %s", addr);
	handle_unknown_message(s);
	return;
  }

  /* default if we don't find anything */
  n = snprintf(buf + sizeof(uint16_t), sizeof(buf) - sizeof(uint16_t), "UNKNOWN");

  enter_critical_section(CRIT_CLIENTS);
  c = all_clients;
  while (c != NULL){
	if (c->our_sockname_valid != 0){
	  s_log(LOG_DEBUG, "Checking port = %d (=%d)", ntohs(c->our_sockname.sin_port), port);
	  if ((memcmp(&output_sockaddr.sin_addr, &c->our_sockname.sin_addr, sizeof(output_sockaddr.sin_addr)) == 0) && (port == ntohs(c->our_sockname.sin_port))) {
		s_log(LOG_DEBUG, "Found");

		control_message = CONTROL_SPLICE;
		
		if (write(c->control_master, &control_message, 1) != 1){
		  s_log(LOG_ERR, "INTERNAL_ERROR: failed to write %d on control connection", CONTROL_SPLICE);
		  break;
		}
		if (write(c->control_master, &receive_fd, sizeof(receive_fd)) != sizeof(receive_fd)) {
		  s_log(LOG_ERR, "INTERNAL_ERROR: failed to write fd %d on control connection", receive_fd);
		  break;
		}
		control_message = CONTROL_NONE;
		read(c->control_master, &control_message, 1);
		if (control_message != CONTROL_SPLICE_DONE){
		  s_log(LOG_ERR, "INTERNAL_ERROR: failed to read %d on control connection", CONTROL_SPLICE_DONE);
		  break;
		}
			
		s_log(LOG_DEBUG, "Received %d", CONTROL_SPLICE_DONE);
		n = snprintf(buf + sizeof(uint16_t), sizeof(buf) - sizeof(uint16_t), "OK");
		break;
	  }
	}
	c = c->next;
  }
  leave_critical_section(CRIT_CLIENTS);

  *((uint16_t*)buf) = n;
  if ((n = send(s, buf, n + sizeof(uint16_t), 0)) < 0){
	s_log(LOG_ERR, "Failed to reply to a message on the control socket: %s", strerror(errno)); 
  }
}


static void handle_control_connection(){
  ssize_t n;
  char buf[1024];
  struct sockaddr_un addr;
  socklen_t addrlen;
  int s;
  int received_fd;

  struct msghdr msg;
  struct iovec vec;
  struct cmsghdr *cmsg;
  char cbuf[CMSG_SPACE(sizeof(s))];

  s = accept(control_socket_fd, (struct sockaddr*)&addr, &addrlen);
  if (s < 0) {
	s_log(LOG_ERR, "Failed to accept a connection on the control socket: %s", strerror(errno));
	return;
  }

  bzero(buf, sizeof(buf));
  msg.msg_name=&addr; /* NULL might be ok */
  msg.msg_namelen=sizeof(addr);
  vec.iov_base=buf;
  vec.iov_len=sizeof(buf);
  msg.msg_iov=&vec;

  msg.msg_iovlen=1;

  msg.msg_control = cbuf;
  msg.msg_controllen = sizeof(cbuf);


  if ((n = recvmsg(s, &msg, 0)) < 0){
	s_log(LOG_ERR, "Failed to read a message on the control socket (n=%d) fd=%d errno=%s", n, control_socket_fd, strerror(errno));
	goto out;
  }
  s_log(LOG_DEBUG, "Received a control message: %s (length %d)", buf + sizeof(uint16_t), n);
  
  if (strcasecmp(buf + sizeof(uint16_t), PING) == 0){
	handle_control_ping(s);
	goto out;
  }
  if (strncasecmp(buf + sizeof(uint16_t), STAT, strlen(STAT)) == 0){
	handle_control_stat(s, buf + sizeof(uint16_t) + strlen(STAT) + 1, n - sizeof(uint16_t) - strlen(STAT) - 1);
	goto out;
  }
  if (strncasecmp(buf + sizeof(uint16_t), SPLICE, strlen(SPLICE)) == 0){
	if (msg.msg_controllen > 0){
	  cmsg = CMSG_FIRSTHDR(&msg);
	  if(cmsg->cmsg_level == SOL_SOCKET && (cmsg->cmsg_type == SCM_RIGHTS)) {
		received_fd=*(int*)CMSG_DATA(cmsg);
		handle_splice(s, buf + sizeof(uint16_t) + strlen(SPLICE) + 1, n - sizeof(uint16_t) - strlen(SPLICE) - 1, received_fd);
		goto out;
 	  }
	}
	s_log(LOG_ERR, "SPLICE requires a file descriptor");
  }


  handle_unknown_message(s);

 out:
  close(s);
}

static void accept_connection(LOCAL_OPTIONS *opt) {
    SOCKADDR_UNION addr;
    char from_address[IPLEN];
    int s;
    socklen_t addrlen;

    addrlen=sizeof(SOCKADDR_UNION);
    while((s=accept(opt->fd, &addr.sa, &addrlen))<0) {
        switch(get_last_socket_error()) {
            case EINTR:
                break; /* retry */
            case EMFILE:
#ifdef ENFILE
            case ENFILE:
#endif
#ifdef ENOBUFS
            case ENOBUFS:
#endif
            case ENOMEM:
                sleep(1); /* temporarily out of resources - short delay */
            default:
                sockerror("accept");
                return; /* error */
        }
    }
    s_ntop(from_address, &addr);
    s_log(LOG_DEBUG, "%s accepted FD=%d from %s",
        opt->servname, s, from_address);
    if(max_clients && num_clients>=max_clients) {
        s_log(LOG_WARNING, "Connection rejected: too many clients (>=%d)",
            max_clients);
        closesocket(s);
        return;
    }
#ifdef FD_CLOEXEC
    fcntl(s, F_SETFD, FD_CLOEXEC); /* close socket in child execvp */
#endif
    if(create_client(opt->fd, s, alloc_client_session(opt, s, s), client)) {
        s_log(LOG_ERR, "Connection rejected: create_client failed");
        closesocket(s);
        return;
    }
    enter_critical_section(CRIT_CLIENTS); /* for multi-cpu machines */
    ++num_clients;
    leave_critical_section(CRIT_CLIENTS);
}

static void get_limits(void) {
#ifdef USE_WIN32
    max_clients=0;
    s_log(LOG_NOTICE, "No limit detected for the number of clients");
#else
    max_fds=0; /* unlimited */

#if defined HAVE_SYSCONF
    max_fds=sysconf(_SC_OPEN_MAX);
    if(max_fds<0)
        ioerror("sysconf");
#elif defined HAVE_GETRLIMIT
    struct rlimit rlim;
    if(getrlimit(RLIMIT_NOFILE, &rlim)<0)
        ioerror("getrlimit");
    else
        max_fds=rlim.rlim_cur;
    if(max_fds==RLIM_INFINITY)
        max_fds=0; /* RLIM_INFINITY should be equal to zero, anyway */
#endif
    s_log(LOG_INFO, "file ulimit = %d%s (can be changed with 'ulimit -n')",
        max_fds, max_fds ? "" : " (unlimited)");
#ifdef USE_POLL
    s_log(LOG_INFO, "poll() used - no FD_SETSIZE limit for file descriptors");
#else
    s_log(LOG_INFO,
        "FD_SETSIZE = %d (some systems allow to increase this value)",
        FD_SETSIZE);
    if(!max_fds || max_fds>FD_SETSIZE)
        max_fds=FD_SETSIZE;
#endif
    if(max_fds && max_fds<16) /* stunnel needs at least 16 file desriptors */
        max_fds=16;
    if(max_fds) {
        max_clients=max_fds>=256 ? max_fds*125/256 : (max_fds-6)/2;
        s_log(LOG_NOTICE, "%d clients allowed", max_clients);
    } else {
        max_clients=0;
        s_log(LOG_NOTICE, "No limit detected for the number of clients");
    }
#endif
}

#if !defined (USE_WIN32) && !defined (__vms)
    /* chroot and set process user and group(s) id */
static void drop_privileges(void) {
    int uid=0, gid=0;
    struct group *gr;
#ifdef HAVE_SETGROUPS
    gid_t gr_list[1];
#endif
    struct passwd *pw;

    /* Get the integer values */
    if(options.setgid_group) {
        gr=getgrnam(options.setgid_group);
        if(gr)
            gid=gr->gr_gid;
        else if(atoi(options.setgid_group)) /* numerical? */
            gid=atoi(options.setgid_group);
        else {
            s_log(LOG_ERR, "Failed to get GID for group %s",
                options.setgid_group);
            exit(1);
        }
    }
    if(options.setuid_user) {
        pw=getpwnam(options.setuid_user);
        if(pw)
            uid=pw->pw_uid;
        else if(atoi(options.setuid_user)) /* numerical? */
            uid=atoi(options.setuid_user);
        else {
            s_log(LOG_ERR, "Failed to get UID for user %s",
                options.setuid_user);
            exit(1);
        }
    }

#ifdef HAVE_CHROOT
    /* chroot */
    if(options.chroot_dir) {
        if(chroot(options.chroot_dir)) {
            sockerror("chroot");
            exit(1);
        }
        if(chdir("/")) {
            sockerror("chdir");
            exit(1);
        }
    }
#endif /* HAVE_CHROOT */

    /* Set uid and gid */
    if(gid) {
        if(setgid(gid)) {
            sockerror("setgid");
            exit(1);
        }
#ifdef HAVE_SETGROUPS
        gr_list[0]=gid;
        if(setgroups(1, gr_list)) {
            sockerror("setgroups");
            exit(1);
        }
#endif
    }
    if(uid) {
        if(setuid(uid)) {
            sockerror("setuid");
            exit(1);
        }
    }
}

static void daemonize(void) { /* go to background */
#if defined(HAVE_DAEMON) && !defined(__BEOS__)
    if(daemon(0,0)==-1) {
        ioerror("daemon");
        exit(1);
    }
#else
    chdir("/");
    switch(fork()) {
    case -1:    /* fork failed */
        ioerror("fork");
        exit(1);
    case 0:     /* child */
        break;
    default:    /* parent */
        exit(0);
    }
    close(0);
    close(1);
    close(2);
#endif
#ifdef HAVE_SETSID
    setsid(); /* Ignore the error */
#endif
}

static void create_pid(void) {
    int pf;
    char pid[STRLEN];

    if(!options.pidfile) {
        s_log(LOG_DEBUG, "No pid file being created");
        return;
    }
    if(options.pidfile[0]!='/') {
        s_log(LOG_ERR, "Pid file (%s) must be full path name", options.pidfile);
        /* Why?  Because we don't want to confuse by
           allowing '.', which would be '/' after
           daemonizing) */
        exit(1);
    }
    options.dpid=(unsigned long)getpid();

    /* silently remove old pid file */
    unlink(options.pidfile);
    if((pf=open(options.pidfile, O_WRONLY|O_CREAT|O_TRUNC|O_EXCL,0644))==-1) {
        s_log(LOG_ERR, "Cannot create pid file %s", options.pidfile);
        ioerror("create");
        exit(1);
    }
    sprintf(pid, "%lu\n", options.dpid);
    write(pf, pid, strlen(pid));
    close(pf);
    s_log(LOG_DEBUG, "Created pid file %s", options.pidfile);
    atexit(delete_pid);
}

static void delete_pid(void) {
    s_log(LOG_DEBUG, "removing pid file %s", options.pidfile);
    if((unsigned long)getpid()!=options.dpid)
        return; /* current process is not main daemon process */
    if(unlink(options.pidfile)<0)
        ioerror(options.pidfile); /* not critical */
}


static void create_control_socket(void) {
	struct sockaddr_un addr;
	
    if(!options.control_socket) {
        s_log(LOG_DEBUG, "No control socket being created");
        return;
    }
    if(options.control_socket[0]!='/') {
        s_log(LOG_ERR, "Control socket (%s) must be full path name", options.control_socket);
        /* Why?  Because we don't want to confuse by
           allowing '.', which would be '/' after
           daemonizing) */
        exit(1);
    }
	unlink(options.control_socket);
	if ((control_socket_fd = socket(AF_UNIX, SOCK_STREAM, 0)) < 0){
	  sockerror("control_socket");
	  exit(1);
	}
	memset(&addr, 0, sizeof(struct sockaddr_un));
	addr.sun_family = AF_UNIX;
	strncpy(addr.sun_path, options.control_socket, sizeof(addr.sun_path) - 1);
	if (bind(control_socket_fd, (struct sockaddr *) &addr, sizeof(struct sockaddr_un)) < 0){
	  sockerror("control_socket bind");
	  exit(1);
	}
	if (listen(control_socket_fd, 5) < 0){
	  sockerror("control_socket listen");
	  exit(1);
	}
	s_log(LOG_DEBUG, "Receiving commands on control socket %s", options.control_socket);
}

static void signal_handler(int sig) { /* signal handler */
    s_log(sig==SIGTERM ? LOG_NOTICE : LOG_ERR,
        "Received signal %d; terminating", sig);
    exit(3);
}
#endif /* !defined USE_WIN32 */

void stunnel_info(int raw) {
    char line[STRLEN];

    sprintf(line, "stunnel " VERSION " on " HOST " with %s",
        SSLeay_version(SSLEAY_VERSION));
    if(raw)
        log_raw("%s", line);
    else
        s_log(LOG_NOTICE, "%s", line);

    safecopy(line, "Threading:");
#ifdef USE_UCONTEXT
    safeconcat(line, "UCONTEXT");
#endif
#ifdef USE_PTHREAD
    safeconcat(line, "PTHREAD");
#endif
#ifdef USE_WIN32
    safeconcat(line, "WIN32");
#endif
#ifdef USE_FORK
    safeconcat(line, "FORK");
#endif

#ifdef HAVE_OSSL_ENGINE_H
    safeconcat(line, " SSL:ENGINE");
#endif

    safeconcat(line, " Sockets:");
#ifdef USE_POLL
    safeconcat(line, "POLL");
#else /* defined(USE_POLL) */
    safeconcat(line, "SELECT");
#endif /* defined(USE_POLL) */
#if defined(USE_WIN32) && !defined(_WIN32_WCE)
    if(s_getaddrinfo)
        safeconcat(line, ",IPv6");
    else
        safeconcat(line, ",IPv4");
#else /* defined(USE_WIN32) */
#if defined(USE_IPv6)
    safeconcat(line, ",IPv6");
#else /* defined(USE_IPv6) */
    safeconcat(line, ",IPv4");
#endif /* defined(USE_IPv6) */
#endif /* defined(USE_WIN32) */

#ifdef USE_LIBWRAP
    safeconcat(line, " Auth:LIBWRAP");
#endif

    if(raw)
        log_raw("%s", line);
    else
        s_log(LOG_NOTICE, "%s", line);
}

/* End of stunnel.c */
